lib/InitModelTypes/InitModelTypes.cpp

```diff
@@ -525,31 +458,26 @@ initModelTypesImpl(FunctionMetadataCache &Cache,
   }
 
   // Insert void types for consistency
-  if (InstType->isVoidTy()) {
-    using model::PrimitiveKind::Values::Void;
-    QualifiedType VoidTy(Model.getPrimitiveType(Void, 0), {});
-    TypeMap.insert({ &I, VoidTy });
-    rc_return VoidTy;
-  }
+  if (InstType->isVoidTy())
+    rc_return model::UpcastableType();
```

Comment:
It should return a void primitive here.

Reply:
Wait, are you sure? As far as I can see, this represents a function-like return value: and those are now encoded with an empty type (to allow them to be omitted from the serialized models).

Reply:
Basically in the old version it was part of the guarantees of `initModelTypes` that for each `llvm::Instruction` with non-aggregate type it was able to compute a valid type.

I see your point about the model having `empty()` to represent the return type of functions that return no value.
But the code you're changing here isn't only triggered for calls to functions that return no value.
Also, the types that we insert in the map in `initModelTypes` are never written in the model, so there's no risk of "polluting" return types with `void` where the model now expects `empty`.

It's not that your proposal dosn't make any sense: we're still computing a type and inserting the entry in the map, even if invalid/empty.
In order to make this properly, we should check all clients of `initModelTypes` to make sure that:
1. they don't assume all the types found in the maps are valid, and they allow for `empty` ones
2. they don't do anything special with `void`
3. if they did anything special with `void` in the old code, they now have do the same with `model::UpcastableType::empty()`

But, even if this is possible, it seems to me a more consistent contract with users to just say: `initModelTypes` returns a map, with a model type in it for each `llvm::Instruction` whose `llvm::Type` is not an aggregate. If we have to special-case `void` it makes the whole thing slightly more surprising, for no actual benefit. It also leaves the `empty()` encoding to signal a different failure mode to users, if it turns out to be necessary in the future.

Reply:
Done, let's hope nothing breaks 🤞.